### PR TITLE
Implement revisit export

### LINKED ISSUE #2341: Implement revisit export
### Description

Generate an export where each row represents a step visit. Each row will include these columns

The usual user information columns
Step Name
Step Entered Time
Step Exited Time
Visit Duration
Visit Counter
Revisit Counter
Previous Step Visited
Steps Visited Since Last Visit

### PR DESCRIPTION
**Pull request by geoffreykwan**
1. Log in as a teacher
1. Open the Classroom Monitor for a run that has branching and students have worked on
1. Go to the "Data Export" view
1. Click "Export Step Visit Events"
1. Click "Show Column Explanations" and make sure the explanations are good
1. Try downloading the export with and without student names
1. Try downloading the export with only certain steps selected
1. Make sure all the visit data is accurate, in particular these columns
Node ID
Step Title
Enter Time
Exit Time
Visit Duration
Visit Counter
Revisit Counter
Previous Node ID
Previous Step Title
Steps Since Last Visit

Closes #2341

### CODE REVIEW

File: src/main/webapp/wise5/classroomMonitor/dataExport/exportVisitsController.ts [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+  createRowWithEmptyCells() {
+    return new Array(this.columnNames.length);
+  }
+
+  getStepsBetweenLastVisit(nodeId: string, previousVisits: any[]) {
+    const steps = [];
+    for (let v = previousVisits.length - 1; v > 0; v--) {
+      const previousVisit = previousVisits[v];
+      const previousNodeId = this.getCellInRow(previousVisit, 'Node ID');
+      if (nodeId === previousNodeId) {
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/classroomMonitor/dataExport/exportVisitsController.ts:320]:
```suggestion
      if (previousNodeId === nodeId) {
```

File: src/main/webapp/wise5/classroomMonitor/dataExport/exportVisitsController.ts [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+      if (this.includeStudentNames) {
+        this.setCellInRow(row, 'Student Name 2', userInfo.users[1].name);
+      }
+    }
+    if (userInfo.users[2] != null) {
+      this.setCellInRow(row, 'WISE ID 3', userInfo.users[2].id);
+      if (this.includeStudentNames) {
+        this.setCellInRow(row, 'Student Name 3', userInfo.users[2].name);
+      }
+    }
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/classroomMonitor/dataExport/exportVisitsController.ts:348]: There's a lot of redundancy here. Can we extract to a function, like
```
function setUserInfo(userIndex: number) {
  if (userInfo.users[userIndex] != null) {
      this.setCellInRow(row, `WISE ID ${userIndex+1}`, userInfo.users[userIndex].id);
      if (this.includeStudentNames) {
        this.setCellInRow(row, `Student Name ${userIndex+1}`, userInfo.users[userIndex].name);
      }
    }
  }
}
```

File: src/main/webapp/wise5/test-unit/controllers/dataExport/exportVisitsController.spec.js [hunk 3]
```diff
@@ -0,0 +1,814 @@
...
+    exportVisitsController.initializeIdToChecked(exportVisitsController.nodes);
+    expect(exportVisitsController.idToChecked['group0']).toEqual(true);
+    expect(exportVisitsController.idToChecked['group1']).toEqual(true);
+    expect(exportVisitsController.idToChecked['node1']).toEqual(true);
+    expect(exportVisitsController.idToChecked['node2']).toEqual(true);
+    expect(exportVisitsController.idToChecked['node3']).toEqual(true);
+    expect(exportVisitsController.idToChecked['node4']).toEqual(true);
+    expect(exportVisitsController.idToChecked['node5']).toEqual(true);
+    expect(exportVisitsController.idToChecked['node6']).toEqual(true);
+    expect(exportVisitsController.idToChecked['node7']).toEqual(true);
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/test-unit/controllers/dataExport/exportVisitsController.spec.js:158]: There's a lot of code here and elsewhere in this file that will be adding to our code complexity, as you can see here: https://codeclimate.com/github/WISE-Community/WISE/pull/2354#issues

How about replacing these calls with a call to one function that takes in an array of ids and an expected value?
```
    expectIdsToChecked(['group0', 'group1', 'node1', 'node2', 'node3', 'node4', 'node5', 'node6', 'node7'], true);
```
The function can be something like this?
```
function expectIdsToChecked(ids, isChecked) {
  for (const id of ids) {
    expect(exportVisitsController.idToChecked[id]).toEqual(isChecked);
  }
}
```

File: src/main/webapp/wise5/classroomMonitor/dataExport/exportController.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    } else if (this.isString(cell)) {
+      csvString = this.wrapInDoubleQuotes(cell);
+    } else {
+      csvString = cell;
+    }
+    if (this.isStringTooLarge(csvString)) {
+      /*
+       * the cell value is larger than the allowable excel cell size so we will display the
+       * string "Data Too Large" instead
+       */
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/classroomMonitor/dataExport/exportController.ts:58]:
```suggestion
```

File: src/main/webapp/wise5/test-unit/controllers/dataExport/exportVisitsController.spec.js [hunk 5]
```diff
@@ -0,0 +1,814 @@
...
+    expect(exportVisitsController.workgroupIdNodeIdToVisitCounter['100-node7']).toEqual(0);
+    expect(exportVisitsController.workgroupIdNodeIdToVisitCounter['101-group0']).toEqual(0);
+    expect(exportVisitsController.workgroupIdNodeIdToVisitCounter['101-group1']).toEqual(0);
+    expect(exportVisitsController.workgroupIdNodeIdToVisitCounter['101-node1']).toEqual(0);
+    expect(exportVisitsController.workgroupIdNodeIdToVisitCounter['101-node2']).toEqual(0);
+    expect(exportVisitsController.workgroupIdNodeIdToVisitCounter['101-node3']).toEqual(0);
+    expect(exportVisitsController.workgroupIdNodeIdToVisitCounter['101-node4']).toEqual(0);
+    expect(exportVisitsController.workgroupIdNodeIdToVisitCounter['101-node5']).toEqual(0);
+    expect(exportVisitsController.workgroupIdNodeIdToVisitCounter['101-node6']).toEqual(0);
+    expect(exportVisitsController.workgroupIdNodeIdToVisitCounter['101-node7']).toEqual(0);
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/test-unit/controllers/dataExport/exportVisitsController.spec.js:223]: Same as above for these lines. Extract to a function that maps over the workgroupIdNodeId.

**Comment by breity** [comment]
Looks great. I was thinking we could add one more column: 'Nodes Since Last Visit' to pair with 'Steps Since Last Visit'.

**Pull request merged**